### PR TITLE
#2033: lb: rewrite and fix bugs in OfflineLB/LBDataRestartReader

### LINKED ISSUE #2033: Simplify and fix OfflineLB
**Describe the bug**

**To Reproduce**
1. Example/test/snippet of code that fails
2. Compiler, platform, libraries
3. Run command: number of processors, threading options, etc.

**Expected behavior**

**Screenshots**

**Platform (please complete the following information):**
 - Supercomputer or desktop [e.g., Trinity, Sierra)
 - Compiler environment [e.g., clang-3.9, gcc8]
 - Libraries [e.g., mpich-2, openmpi)
 - Commit ID or release of VT [e.g., vt-1.0.0]

**Additional context**

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #2033

### CODE REVIEW

**Comment by lifflander** [comment]
Tests are coming.

File: src/vt/vrt/collection/balance/lb_data_restart_reader.cc [hunk 1]
```diff
@@ -78,45 +78,15 @@ void LBDataRestartReader::startup() {
...
-
-std::deque<std::set<ElementIDType>> LBDataRestartReader::readIntoElementHistory(
-  LBDataHolder const& lbdh
-) {
-  std::deque<std::set<ElementIDType>> element_history;
-  for (PhaseType phase = 0; phase < lbdh.node_data_.size(); phase++) {
-    std::set<ElementIDType> buffer;
+void LBDataRestartReader::readHistory(LBDataHolder const& lbdh) {
+  num_phases_ = lbdh.node_data_.size();
+  for (PhaseType phase = 0; phase < num_phases_; phase++) {
```
Comment (nlslatt) [inline src/vt/vrt/collection/balance/lb_data_restart_reader.cc:83]: This seems to assume that a file with `N` phases will contain specifically phases `0` through `N-1`. Now that we have the ability to selectively output phases to the LB data files, I think this is no longer a safe assumption.

Do we want to handle cases where the file contains gaps in phase ranges that may not even be reached by this run? We should at least give a better error message than just an out of range exception when we fail to find an expected phase.

Do we want to allow cases where `OfflineLB` is only used for a subset of particularly challenging phases, eliminating the requirement to have a continuous range of phases of history available?

Do we want to allow reading in only a subset of phases from potentially very large files?

File: tests/unit/lb/test_offlinelb.cc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#include <vt/transport.h>
+
```
Comment (nlslatt) [inline tests/unit/lb/test_offlinelb.cc:45]: Looks like you need an additional include.

**Review by nlslatt — COMMENTED** [review]
I wonder if we need to make this more general. Please see the comment below.

File: src/vt/vrt/collection/balance/lb_data_restart_reader.cc (hunk 1 above)
Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_data_restart_reader.cc:83]: Good point. I think you are right that we need to consider cases where not all phases are output.

File: src/vt/vrt/collection/balance/lb_data_restart_reader.cc [hunk 3]
```diff
@@ -78,45 +78,22 @@ void LBDataRestartReader::startup() {
...
-    for (auto const& obj : lbdh.node_data_.at(phase)) {
-      if (obj.first.isMigratable()) {
-        buffer.insert(obj.first.id);
+void LBDataRestartReader::readHistory(LBDataHolder const& lbdh) {
+  num_phases_ = lbdh.node_data_.size();
+  PhaseType prev_known_phase = 0;
+  for (PhaseType phase = 0; phase < num_phases_; phase++) {
+    auto iter = lbdh.node_data_.find(phase);
+    if (iter == lbdh.node_data_.end()) {
+      history_[phase] = history_[prev_known_phase];
```
Comment (nlslatt) [inline src/vt/vrt/collection/balance/lb_data_restart_reader.cc:87]: This is an interesting solution to the problem. Could we end up using too much memory this way? My original thought was that the LB interval would likely not be 1, so only a subset of the phases would be needed anyway. I think it's also possible to mix `OfflineLB` with another load balancer (on different phases) by using an LB config file.

Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_data_restart_reader.cc:87]: I agree that it uses more memory than necessary. It was just the easiest way to write the code. I can do a search each time instead that will find the last previous known phase.

Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_data_restart_reader.cc]: Ok based on our conversation, for the sake of this PR, we are going to assume that the phase specification is dense is fully specified for every phase up to num_phases.

File: src/vt/vrt/collection/balance/lb_data_restart_reader.h [hunk 4]
```diff
@@ -83,67 +83,96 @@ struct LBDataRestartReader : runtime::component::Component<LBDataRestartReader>
...
   template <typename Serializer>
   void serialize(Serializer& s) {
-    s | msgsReceived
-      | totalMove
-      | proxy_
-      | proc_move_list_
-      | proc_phase_runs_LB_;
+    s | proxy_
+      | changed_distro_
+      | num_phases_;
```
Comment (nlslatt) [inline src/vt/vrt/collection/balance/lb_data_restart_reader.h:122]: Should the serializer include `history_`?

Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_data_restart_reader.h]: Fixes

**Review by nlslatt — APPROVED** [review]
Looks good to me

**Pull request merged**